[Codegen 94]  move `extendsForProp` into `parsers-commons`

siddarthkay opened the pull request:
## Summary:

[Codegen 94] This PR attempts to extracts the logic of `extendsForProp` function from the following locations : 
- `parsers/flow/components/extends.js`
- `parsers/typescript/components/props.js`
since they are the same and move the function to `parsers/parsers-commons.js` as requested on https://github.com/facebook/react-native/issues/34872

## Changelog:

[Internal] [Changed] - Move `extendsForProp` to parser-commons and update usages.

## Test Plan:

Run `yarn jest react-native-codegen` and ensure CI is green

siddarthkay commented:
@cipolleschi : This PR is ready for review.

cipolleschi commented on packages/react-native-codegen/src/parsers/parsers-commons.js:
```diff
@@ -817,6 +825,36 @@ function propertyNames(
+  types: TypeDeclarationMap,
+  propName: 'argument' | 'expression',
+): ExtendedPropResult {
+  if (!prop[propName]) {
```
The way we are trying to handle these is by using the `Parser` object.

The [`Parser`](https://github.com/facebook/react-native/blob/main/packages/react-native-codegen/src/parsers/parser.js) object is an abstract class which is implemented by a [`TypeScriptParser`](https://github.com/facebook/react-native/blob/main/packages/react-native-codegen/src/parsers/typescript/parser.js) and a [`FlowParser`](https://github.com/facebook/react-native/blob/main/packages/react-native-codegen/src/parsers/flow/parser.js).

So, what we should do is to extends the `Parser` class so that it has a function like:
```ts
argumentForProp(prop: PropAST): $FlowFixMe; // <-- You may have to use $FlowFixMe or define what a PropAST is
```

And implement the function in the two parsers
```ts
// In the FlowParser
argumentForProp(prop: PropAST): $FlowFixMe {
  return prop.argument
}

// In the TypeScriptParser
argumentForProp(prop: PropAST): $FlowFixMe {
  return prop.expression
}
```

siddarthkay replied:
Oh yes, This is much neater! Thanks for explaining @cipolleschi

cipolleschi commented on packages/react-native-codegen/src/parsers/parsers-commons.js:
```diff
@@ -817,6 +825,36 @@ function propertyNames(
+    console.log('null', prop);
+  }
+
+  const name = prop[propName]?.id
+    ? prop[propName].id.name
+    : prop[propName]?.name;
```
Similar concern as above, we should extract this in a `nameForArgument` function in the parsers.
It could be that the parsers already have this method implemented.

cipolleschi commented on packages/react-native-codegen/src/parsers/typescript/components/props.js:
```diff
@@ -96,7 +74,7 @@ function getProps(
     // find extends
     if (prop.type === 'TSExpressionWithTypeArguments') {
-      const extend = extendsForProp(prop, types);
+      const extend = extendsForProp(prop, types, 'expression');
```
We should pass a `Parser` object here, not an `'expression'`.

cipolleschi commented on packages/react-native-codegen/src/parsers/flow/components/extends.js:
```diff
@@ -56,7 +34,7 @@ function getExtendsProps(
   return typeDefinition
     .filter(prop => prop.type === 'ObjectTypeSpreadProperty')
-    .map(prop => extendsForProp(prop, types))
+    .map(prop => extendsForProp(prop, types, 'argument'))
```
We should pass a `Parser` object here, not an `'argument'`.

See below.

cipolleschi reviewed (changes requested):
Hi @siddarthkay, thank you so much for taking the time to work on this!

I left some comments to improve the current implementation, following a pattern we have established in other Codegen tasks.

Let me know if it is clear and/or if you need help! :D

siddarthkay commented:
Hi @cipolleschi : I've addressed your feedback and I also took the liberty of removing `type PropAST = Object;` from 5 files and moving it to `packages/react-native-codegen/src/parsers/utils.js` hope this is okay?

Let me know :)

cipolleschi reviewed (approved):
Good! :D Thanks for the changes!

siddarthkay commented:
Hi @cipolleschi : I've rebased my branch and resolved all conflicts,
Would you mind taking a look at this PR again and let me what are the next steps here?
I'm eager and excited to get my 1st PR merged into react-native 👍🏻

rshest commented:
Hey @siddarthkay, sorry for back and forth, but this change severely conflicts with https://github.com/facebook/react-native/pull/36891, which has been now merged to master (it merges together `getExtendedProps` and `getProps`).

Could you please rebase this PR to master and try and resolve the merge conflicts? 

Thanks!!!

rshest reviewed (changes requested):
Please see my comment above, this requires a rebase to master and merge/resolve conflicts.

siddarthkay commented:
Hi @rshest : I think I already resolved those conflicts few hours ago, I have just now rebased to latest `main` branch 
and tests seem to pass. 
Could you please check again?

siddarthkay commented:
I see `test_android_template-NewArch-Debug-Hermes` failed on CI with :

```log
error An unexpected error occurred: "http://localhost:4873/yallist/-/yallist-3.1.1.tgz: 
Request failed \"500 Internal Server Error\""

error Command failed with exit code 1: yarn add /root/react-native/packages/react-native
Error: Command failed with exit code 1: yarn add /root/react-native/packages/react-native
    at makeError (/root/react-native/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/root/react-native/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async createFromTemplate (/root/react-native/node_modules/@react-native-community/cli/build/commands/init/init.js:96:5)
    at async Object.initialize [as func] (/root/react-native/node_modules/@react-native-community/cli/build/commands/init/init.js:179:3)
    at async Command.handleAction (/root/react-native/node_modules/@react-native-community/
    cli/build/index.js:109:9)
node:child_process:935
    throw err;
    ^

Error: Command failed: node cli.js init AndroidTemplateProject --directory 
/tmp/AndroidTemplateProject --template /root/react-native/packages/react-native 
--verbose --skip-install
    at checkExecSyncError (node:child_process:861:11)
    at execSync (node:child_process:932:15)
    at install (/root/react-native/scripts/template/initialize.js:72:3)
    at Object.<anonymous> (/root/react-native/scripts/template/initialize.js:95:1)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)
    at Module.load (node:internal/modules/cjs/loader:1074:32)
    at Function.Module._load (node:internal/modules/cjs/loader:909:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 521,
  stdout: null,
  stderr: null
}

```

Should I rerun this job on CircleCI : https://app.circleci.com/pipelines/github/facebook/react-native/22620/workflows/be7cfad0-d96d-4c5c-b552-5006294413ff/jobs/580151 ?
OR 
Amend and force push last commit to run all jobs?

Pranav-yadav commented:
@siddarthkay as the error says:
'Request failed \"500 Internal Server Error\"'
It's may be a transient error, due to network or something else, so no need to worry about it. 👍

PS. Reviewer(s) will re-rerun it if required.

siddarthkay commented:
Hi @Pranav-yadav Thanks for explaining!

rshest commented on packages/react-native-codegen/src/parsers/flow/components/props.js:
```diff
@@ -58,11 +56,13 @@ function buildPropSchema(
 ): ExtendsForProp {
-  if (!prop.argument) {
+  const argument = parser.argumentForProp(prop);
+  if (argument) {
```
This now does exactly the opposite of what it used/supposed to do :)

It should instead be:
```
if (!argument) {
```

Otherwise this would break codegen completely, at least on our internal tests.

I am surprised, though, that CI on Github didn't catch this (the only failed test seems inside the app template and is unrelated).

I think we need a follow-up on that, CC @cipolleschi 

@siddarthkay - note that I fixed this internally myself this time, to save one extra back and forth, but normally that would be your responsibility :)

Pranav-yadav replied:
>I am surprised, though, that CI on Github didn't catch this (the only failed test seems inside the app template and is unrelated).

I think, those files don't have tests, but, other tests should have been failed; or may be the current test-suite _doesn't_ test this specific case.
